File: 이진영/프로그래머스/가장 많이 받은 선물.js
function solution(friends, gifts) {
    const N = friends.length;
    let friends_obj = {}
    friends.forEach((v, i) => {
        friends_obj[v] = i;
    })
    
    let scores = Array(N).fill(0);
    let table = Array.from(Array(N), () => Array(N).fill(0));
    
    gifts.forEach(g => {
        const [from, to] = g.split(" ").map(v => +friends_obj[v]);
        scores[from]++;
        scores[to]--;
        table[from][to]++;
    })
    
    let next_week = Array(N).fill(0);
    
    for (let i = 0; i < N; i++) {
        for (let j = i + 1; j < N; j++) {
            if (table[i][j] > table[j][i]) {
                next_week[i]++;
            } else if (table[i][j] < table[j][i]) {
                next_week[j]++;
            } else {
                if (scores[i] !== scores[j]) {
                    next_week[scores[i] > scores[j] ? i : j]++;
                }
            }
        }
    }
    
    return Math.max(...next_week);
}

File: 김채린/추가문제/프로그래머스/가장 많이 받은 선물.java
import java.util.*;
class Solution {
    public int solution(String[] friends, String[] gifts) {
        int[] result = new int[friends.length];
        int[][] arr = new int[friends.length][friends.length];
        HashMap<String, Integer> index = new HashMap<>();
        int[] giftCount = new int[friends.length];
        for(int i=0; i<friends.length; i++){
            index.put(friends[i],i);
        }
        for(int i=0; i<gifts.length; i++){
            String[] now = gifts[i].split(" ");
            int sender =index.get(now[0]);
            int recipient =index.get(now[1]);
            giftCount[sender]++;
            giftCount[recipient]--;
            arr[sender][recipient]++;
        }
        String aaaa = "";
        int maxCount = 0;
        for(int i=0; i<friends.length; i++){
            for(int j=i+1; j<friends.length; j++){
                int first = arr[i][j];
                int second = arr[j][i];
                if(first>second){
                    result[i]++;
                }else if(first<second){
                    result[j]++;
                }else{
                    first = giftCount[i];
                    second =giftCount[j];
                    if(first>second){
                        result[i]++;
                    }else if(first<second){
                        result[j]++;
                    }
                }
            }
            maxCount = Math.max(maxCount,result[i]);
        }
        return maxCount;
    }
}
